### ISSUE TITLE
Scoped storage: allow keeping collection in the legacy location

### ISSUE THREAD

**Issue by oakkitten** [ankidroid/Anki-Android#13134]
You didn't want to think about this but here you are 😛

...at some time I suggested allowing importing from an existing AnkiDroid folder. Why? Imagine that instead of updating the app, which would retain access to legacy storage, the user uninstalls & reinstalls AnkiDroid, which would make it lose that access. Wot do? This is what the dreary Figma points are about, such as “your current storage location ... will be inaccessible if the app is unistalled”. Importing from a folder is supposed to be easy to implement, so why not?

But then I thought that I would honestly keep my AnkiDroid folder where it is. I hardly care about speed. I can wait for sync. Reviewing is fast enough. I care more about the fact that the folder doesn't disappear by any chance. Perhaps many an user would feel likewise?

So I suggest completely reworking scoped storage with the idea that we can keep the data where it is. We'd need to ask the user to grant the RW permission to the AnkiDroid folder, a dialog that many users will be familiar to, and easy enough to comprehend for those who never saw it.

**Comment by david-allison** [comment]
1) We can do this with `MANAGE_EXTERNAL_STORAGE`. We can do this for F-Droid and apk downloads via GitHub, but we won't get the permission on Google Play.

I'd be fine with this

2) I'd support a PR which makes AnkiDroid work using `DocumentFile`/`ACTION_OPEN_DOCUMENT_TREE` which would work with Google Play. 

I strongly suspect that it's infeasible as `DocumentFile` isn't a `java.io.File`, so we wouldn't be able to open the collection (for example) : https://stackoverflow.com/questions/70166725/opening-large-sqlite-databases-on-android-11 

There's a lot of jank in converting from `DocumentFile` to `File`. I suspect that's also infeasible.

> `Please notice that accessing files with [File] only works on app private directory since Android 10. You had better to stay using [DocumentFile].`

https://github.com/anggrayudi/SimpleStorage/blob/ec25051a0e76c19b5e78f29f3405526c3566bce9/storage/src/main/java/com/anggrayudi/storage/file/DocumentFileExt.kt#L247-L268

**Comment by oakkitten** [comment]
As I understand, we currently have no problem with Android 10 and can continue doing what we are doing now. Android 11+ [seems to be](https://developer.android.com/training/data-storage/shared/documents-files#retrieve-equivalent-media-uri:~:text=Using%20the%20media%20store%20URI%2C%20you%20can%20more%20easily%20access%20media%20files%20from%20shared%20storage) allowing using [`File` API and native file access](https://developer.android.com/training/data-storage/shared/media#direct-file-paths) for folders that we have permissions for.

**Comment by Arthur-Milchior** [comment]
I'm sorry @oakkitten but i don't understand what you mean by File API here.
As far as I understand the examples you give, it shows that you can access the file content; which is clearly a good thing. But I fear it does not help with the underlying SQLite issue. That is, even if we access the file bytestream, we can't provide it as-is to sqlite. Sqlite still wants the path of the database, and we still have no reliable way to get it.

The best I can imagine is that we ask the user to both:
* give us permission through  ACTION_OPEN_DOCUMENT_TREE
* to manually enter the path, (unless it's `~/AnkiDroid`, that we can check by default).

I'd find it ugly to decouple getting the permission and knowing the name of the folder, but I guess it can also be the most realistic solution we have.

**Comment by oakkitten** [comment]
If you follow the link, you can see more details. `File` API is just `java.io.File`, which is to say, you should be able to access the files like you normally would—provided that we can figure out how to get a physical path out of whatever `ACTION_OPEN_DOCUMENT_TREE` gives us.

To be completely clear, this is how I imagine this to work, roughly:
* When AnkiDroid is started, it tries to open the collection without asking for any permission
* If that fails,
  * On Android ≤ 10, AnkiDroid asks for permission to access all files, like it currently does
  * On Android ≥ 11, AnkiDroid asks for access to a specific folder, stores the new path and uses that

**Comment by oakkitten** [comment]
Okay, after some investigation I have to conclude that this is not possible. While the above link mentions that Android 11 brings back “direct file paths”, these seem to be limited to files that you get from media store, that is, pictures or video files. This is what documentation says; I was just (really) hoping that it was slightly incorrect, and that you could natively access all files that you have permissions for, since, well, the implementation is already there. Oh well. 

![image](https://user-images.githubusercontent.com/1710718/214105235-7db5a2d5-cf06-48e7-b396-0357066099bc.png)

I'm closing this now, but I want to note that while we cannot have `File` API access to files that we get using `ACTION_OPEN_DOCUMENT_TREE`, we can still access them through SAF. This can at least allow us to *import* (migrate) collections from legacy folders. However this is a different issue. And I'm not sure if it's worth going this way because of the complexity of the task—it would mean remaking a large portion of the migration code.

**Comment by david-allison** [comment]
I've recently been introduced to [FSAF](https://github.com/K1rakishou/Fuck-Storage-Access-Framework) which would work as an abstraction layer. 

I'm semi-open to a discussion on using it for a migration from a selected folder, but to be frank, I have zero trust in Android's `DocumentFile` abstraction, and I don't suspect its flaws can be patched over.

**Comment by Arthur-Milchior** [comment]
Can you please document what investigation you did, so that other don't do the same?
I thought your idea was great. I also wanted to test it, and failed.I suspect we may have tried the same thing here.

Using #13149 from @ParthTagalpallewar , I tested giving myself permission to access ~/AnkiDroid, and then changed the Collection folder to it. I was oping to be able to open it. Sadly, as you probably find yourself, there is no write access to the folder /storage/emulated/0/AnkiDroid. Too bad, it was a great idea.

I reached the same conclusion than @oakkitten, it means that we'll need to use DocumentFile if we want to do migration of legacy folder using a newly installed version of the app. Which is why I created #13145. 
I think it would be nice to have. Ideally on the first release on stable, to be sure that even if someone tries to uninstall and reinstall, they can access their collection. However, I can also understand it would not be a release blocker.

**Comment by oakkitten** [comment]
>Can you please document what investigation you did, so that other don't do the same?

What I did was I made AnkiDroid fire an `ACTION_OPEN_DOCUMENT_TREE` Intent, and after getting access to `content://com.android.externalstorage.documents/tree/primary:AnkiDroid`, tried reading from `/storage/emulated/0/AnkiDroid` direcly, which failed. I then tried various URI conversions to see if there are any tricks or hidden paths that can be used alternatively, all attempts failing with errors like `java.lang.UnsupportedOperationException: Unsupported Uri content://com.android.externalstorage.documents/tree/primary%3AAnkiDroid`. (Also to find errors that can lead to useful SO questions.) Finally, I checked SO for all relevant information and didn't find one hint saying that what we wanted was possible.
